File: plugin-stylejson/src/main/java/com/mapbox/pluginstylejson/StyleJsonParser.java
package com.mapbox.pluginstylejson;

import android.support.annotation.NonNull;

import com.google.gson.Gson;
import com.google.gson.JsonObject;
import com.mapbox.geojson.Feature;
import com.mapbox.geojson.FeatureCollection;
import com.mapbox.geojson.GeometryCollection;
import com.mapbox.geojson.LineString;
import com.mapbox.geojson.MultiLineString;
import com.mapbox.geojson.MultiPoint;
import com.mapbox.geojson.MultiPolygon;
import com.mapbox.geojson.Point;
import com.mapbox.geojson.Polygon;
import com.mapbox.mapboxsdk.style.sources.GeoJsonSource;

public class StyleJsonParser {
  private Gson gson = new Gson();

  GeoJsonSource parserGeoJsonSource(@NonNull String id, @NonNull String json) {
    GeoJsonSource geoJsonSource = new GeoJsonSource(id);
    JsonObject source = gson.fromJson(json, JsonObject.class);
    if (source.has("data")) {
      JsonObject data = source.getAsJsonObject("data");
      String dataString = data.getAsString();
      if (dataString.startsWith("http")) {
        geoJsonSource.setUrl(dataString);
      } else {
        String type = data.get("type").getAsString();
        if ("Feature".equals(type)) {
          geoJsonSource.setGeoJson(Feature.fromJson(dataString));
        } else if ("FeatureCollection".equals(type)) {
          geoJsonSource.setGeoJson(FeatureCollection.fromJson(dataString));
        } else if ("GeometryCollection".equals(type)) {
          geoJsonSource.setGeoJson(GeometryCollection.fromJson(dataString));
        } else if ("LineString".equals(type)) {
          geoJsonSource.setGeoJson(LineString.fromJson(dataString));
        } else if ("MultiLineString".equals(type)) {
          geoJsonSource.setGeoJson(MultiLineString.fromJson(dataString));
        } else if ("MultiPoint".equals(type)) {
          geoJsonSource.setGeoJson(MultiPoint.fromJson(dataString));
        } else if ("MultiPolygon".equals(type)) {
          geoJsonSource.setGeoJson(MultiPolygon.fromJson(dataString));
        } else if ("Point".equals(type)) {
          geoJsonSource.setGeoJson(Point.fromJson(dataString));
        } else if ("Polygon".equals(type)) {
          geoJsonSource.setGeoJson(Polygon.fromJson(dataString));
        }
      }
    }

    return geoJsonSource;
  }
}

File: plugin-stylejson/src/main/java/com/mapbox/pluginstylejson/StyleJsonParser.kt
package com.mapbox.pluginstylejson

import com.google.gson.Gson
import com.google.gson.JsonObject
import com.mapbox.geojson.*
import com.mapbox.mapboxsdk.style.sources.GeoJsonSource
import com.mapbox.mapboxsdk.style.sources.VectorSource
import com.mapbox.pluginstylejson.StyleJsonContants.GEO_JSON_FEATURE
import com.mapbox.pluginstylejson.StyleJsonContants.GEO_JSON_FEATURE_COLLECTION
import com.mapbox.pluginstylejson.StyleJsonContants.GEO_JSON_GEOMETRY_COLLECTION
import com.mapbox.pluginstylejson.StyleJsonContants.GEO_JSON_LINE_STRING
import com.mapbox.pluginstylejson.StyleJsonContants.GEO_JSON_MULTI_POINT
import com.mapbox.pluginstylejson.StyleJsonContants.GEO_JSON_MULTI_POLYGON
import com.mapbox.pluginstylejson.StyleJsonContants.GEO_JSON_NULTI_LINE_STRING
import com.mapbox.pluginstylejson.StyleJsonContants.GEO_JSON_POINT
import com.mapbox.pluginstylejson.StyleJsonContants.GEO_JSON_POLYGON

class StyleJsonParser {
      private val gson = Gson()

    internal fun parserGeoJsonSource(id: String, json: String): GeoJsonSource {
        val geoJsonSource = GeoJsonSource(id)
        val source = gson.fromJson<JsonObject>(json, JsonObject::class.java)
        if (source.has("data")) {
            val data = source.getAsJsonObject("data")
            val dataString = data.asString
            if (dataString.startsWith("http")) {
                geoJsonSource.url = dataString
            } else {
                val type = data.get("type").asString
                when (type) {
                    GEO_JSON_FEATURE -> geoJsonSource.setGeoJson(Feature.fromJson(dataString))
                    GEO_JSON_FEATURE_COLLECTION -> geoJsonSource.setGeoJson(FeatureCollection.fromJson(dataString))
                    GEO_JSON_GEOMETRY_COLLECTION -> geoJsonSource.setGeoJson(GeometryCollection.fromJson(dataString))
                    GEO_JSON_LINE_STRING -> geoJsonSource.setGeoJson(LineString.fromJson(dataString))
                    GEO_JSON_NULTI_LINE_STRING -> geoJsonSource.setGeoJson(MultiLineString.fromJson(dataString))
                    GEO_JSON_MULTI_POINT -> geoJsonSource.setGeoJson(MultiPoint.fromJson(dataString))
                    GEO_JSON_MULTI_POLYGON -> geoJsonSource.setGeoJson(MultiPolygon.fromJson(dataString))
                    GEO_JSON_POINT -> geoJsonSource.setGeoJson(Point.fromJson(dataString))
                    GEO_JSON_POLYGON -> geoJsonSource.setGeoJson(Polygon.fromJson(dataString))
                }
            }
        }

        return geoJsonSource
    }

    internal fun parserVectorSource(id: String, json: String): VectorSource {

    }
}
